# Robustness Checks and Presentation Notes

This notebook runs the companion robustness pipeline for the damped shock model and keeps a short, presentation-ready record of the main messages.

What it does:
- runs the scenario summary pipeline
- runs the damping-scale sensitivity sweep
- runs the bootstrap perturbation check
- stores compact outputs in `outputs/` and `outputs/figures/`

In [1]:
from pathlib import Path
import sys

import numpy as np
import pandas as pd
from IPython.display import display

ROOT_DIR = Path.cwd().resolve()
if ROOT_DIR.name == 'notebooks':
    ROOT_DIR = ROOT_DIR.parent

if str(ROOT_DIR) not in sys.path:
    sys.path.insert(0, str(ROOT_DIR))

from robustness_checks import (
    ensure_output_dirs,
    build_base_model,
    select_reference_shocks,
    run_scenarios,
    run_scale_sweep,
    run_bootstrap,
)

pd.set_option('display.max_rows', 20)
pd.set_option('display.max_columns', 20)
np.set_printoptions(suppress=True, precision=4)

output_dir, figure_dir = ensure_output_dirs()
print(f'Outputs will be written to: {output_dir}')
print(f'Figures will be written to: {figure_dir}')

Outputs will be written to: C:\Users\Aditya\Documents\wion-shock-model\outputs
Figures will be written to: C:\Users\Aditya\Documents\wion-shock-model\outputs\figures


In [2]:
# Robustness run settings
year = 2018
n_iter = 10
scales = [0.5, 0.75, 1.0, 1.25, 1.5]
n_bootstrap = 5
bootstrap_sigma = 0.05
seed = 42

base = build_base_model(year=year, n_iter=n_iter)
shocks = select_reference_shocks(base['A'], base['V'], base['d'], base['phi_node'])

all_results, scenario_summary = run_scenarios(base['sim'], shocks)
scale_summary = run_scale_sweep(base, shocks, scales, output_dir, figure_dir)
bootstrap_draws, bootstrap_summary = run_bootstrap(
    base=base,
    shocks=shocks,
    n_boot=n_bootstrap,
    sigma=bootstrap_sigma,
    seed=seed,
    output_dir=output_dir,
    figure_dir=figure_dir,
)

presentation_summary = pd.DataFrame({
    'scenario': scenario_summary['label'],
    'mean_delta_rl': scenario_summary['mean_delta_rl'],
    'top_country': scenario_summary['top_country_delta'],
    'top_country_delta': scenario_summary['top_country_delta_value'],
})

display(scenario_summary)
display(scale_summary.head(10))
display(bootstrap_summary)
display(presentation_summary)

Active nodes after zero-output removal: 3928
Computing V_j: Upstream Systemic Exposure
Method: Option C — Global Export Concentration (IRR)

[1/3] Building input distribution matrix...
  Input distribution : (3928, 3928)
  Zero-input sectors : 0

[2/3] Computing IRR: global export concentration per sector...
  Computing export flows (cross-border only)...
  This may take a moment for 3928 nodes...
  Nodes with positive exports : 3918
  Non-exporting nodes         : 10
  Total export volume         : 12398232.52

  Global EXPORT concentration IRR by sector code:
  Top 10 most concentrated (irreplaceable export markets):
sector_code
B05         0.222951
C13T15      0.182886
B06         0.154113
B09         0.145945
C302T309    0.140728
C301        0.136797
J58T60      0.119016
C26         0.117303
H53         0.109292
C27         0.106891

  Bottom 10 least concentrated (substitutable export markets):
sector_code
H52       0.030460
C10T12    0.034768
H49       0.036281
F         0.045223

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


  LPI countries not mapped to ICIO (78): ['Qatar', 'Panama', 'Oman', 'Rwanda', 'Bahrain', 'Kuwait', 'Serbia', 'Kenya', 'Bosnia and Herzegovina', 'Paraguay', 'Benin', 'Montenegro', 'Mauritius', 'Lebanon', 'Macedonia, FYR', 'Uruguay', 'Maldives', 'Dominican Republic', 'Albania', 'Djibouti', 'Burkina Faso', 'Armenia', 'Honduras', 'Sri Lanka', 'Mali', 'Malawi', 'Uzbekistan', 'El Salvador', 'Uganda', 'Solomon Islands', 'Ghana', 'Comoros', 'Kyrgyz Republic', 'Zambia', 'Bahamas, The', 'Jamaica', 'Nepal', 'Congo, Rep.', 'Moldova', 'Algeria', 'Togo', 'Georgia', 'Sudan', 'Chad', 'Trinidad and Tobago', 'Guatemala', 'Turkmenistan', 'Gambia, The', 'Madagascar', 'Guinea-Bissau', 'Mongolia', 'Bolivia', 'Guyana', 'Fiji', 'Tajikistan', 'Mauritania', 'Equatorial Guinea', 'Syrian Arab Republic', 'Lesotho', 'Yemen, Rep.', 'Venezuela, RB', 'Liberia', 'Somalia', 'Guinea', 'Cuba', 'Iraq', 'Papua New Guinea', 'Bhutan', 'Gabon', 'Central African Republic', 'Zimbabwe', 'Haiti', 'Libya', 'Eritrea', 'Sierra Leone

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


Damped model ready: A(I-D) computed
  Mean diagonal of D : 0.3588

Shock   : max_V
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Damped  — nodes RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Delta   — mean absorption: 0.0000 | Max absorption: 0.0045 (TUN_O)
Damped* — common-ref RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Delta*  — common-ref mean absorption: 0.0000 | Max absorption: 0.0045 (TUN_O)

Shock   : max_d
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 60 | Max RL: 1.0000 (DNK_H50)
Damped  — nodes RL>10%: 11 | Max RL: 1.0000 (DNK_H50)
Delta   — mean absorption: 0.0121 | Max absorption: 0.4510 (DNK_C301)
Damped* — common-ref RL>10%: 11 | Max RL: 1.0000 (DNK_H50)
Delta*  — common-ref mean absorption: 0.0121 | Max absorption: 0.4510 (DNK_C301)

Shock   : max_phi
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 1 | Max RL: 1.0000 (DEU_A01)
Damped  — nodes RL>10%: 1 | Max RL: 1.0000 (DEU_A01)
Delta   — mean absorption: 0.0021 | Max absorption: 

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


  LPI countries not mapped to ICIO (78): ['Qatar', 'Panama', 'Oman', 'Rwanda', 'Bahrain', 'Kuwait', 'Serbia', 'Kenya', 'Bosnia and Herzegovina', 'Paraguay', 'Benin', 'Montenegro', 'Mauritius', 'Lebanon', 'Macedonia, FYR', 'Uruguay', 'Maldives', 'Dominican Republic', 'Albania', 'Djibouti', 'Burkina Faso', 'Armenia', 'Honduras', 'Sri Lanka', 'Mali', 'Malawi', 'Uzbekistan', 'El Salvador', 'Uganda', 'Solomon Islands', 'Ghana', 'Comoros', 'Kyrgyz Republic', 'Zambia', 'Bahamas, The', 'Jamaica', 'Nepal', 'Congo, Rep.', 'Moldova', 'Algeria', 'Togo', 'Georgia', 'Sudan', 'Chad', 'Trinidad and Tobago', 'Guatemala', 'Turkmenistan', 'Gambia, The', 'Madagascar', 'Guinea-Bissau', 'Mongolia', 'Bolivia', 'Guyana', 'Fiji', 'Tajikistan', 'Mauritania', 'Equatorial Guinea', 'Syrian Arab Republic', 'Lesotho', 'Yemen, Rep.', 'Venezuela, RB', 'Liberia', 'Somalia', 'Guinea', 'Cuba', 'Iraq', 'Papua New Guinea', 'Bhutan', 'Gabon', 'Central African Republic', 'Zimbabwe', 'Haiti', 'Libya', 'Eritrea', 'Sierra Leone

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


  LPI countries not mapped to ICIO (78): ['Qatar', 'Panama', 'Oman', 'Rwanda', 'Bahrain', 'Kuwait', 'Serbia', 'Kenya', 'Bosnia and Herzegovina', 'Paraguay', 'Benin', 'Montenegro', 'Mauritius', 'Lebanon', 'Macedonia, FYR', 'Uruguay', 'Maldives', 'Dominican Republic', 'Albania', 'Djibouti', 'Burkina Faso', 'Armenia', 'Honduras', 'Sri Lanka', 'Mali', 'Malawi', 'Uzbekistan', 'El Salvador', 'Uganda', 'Solomon Islands', 'Ghana', 'Comoros', 'Kyrgyz Republic', 'Zambia', 'Bahamas, The', 'Jamaica', 'Nepal', 'Congo, Rep.', 'Moldova', 'Algeria', 'Togo', 'Georgia', 'Sudan', 'Chad', 'Trinidad and Tobago', 'Guatemala', 'Turkmenistan', 'Gambia, The', 'Madagascar', 'Guinea-Bissau', 'Mongolia', 'Bolivia', 'Guyana', 'Fiji', 'Tajikistan', 'Mauritania', 'Equatorial Guinea', 'Syrian Arab Republic', 'Lesotho', 'Yemen, Rep.', 'Venezuela, RB', 'Liberia', 'Somalia', 'Guinea', 'Cuba', 'Iraq', 'Papua New Guinea', 'Bhutan', 'Gabon', 'Central African Republic', 'Zimbabwe', 'Haiti', 'Libya', 'Eritrea', 'Sierra Leone

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


  LPI countries not mapped to ICIO (78): ['Qatar', 'Panama', 'Oman', 'Rwanda', 'Bahrain', 'Kuwait', 'Serbia', 'Kenya', 'Bosnia and Herzegovina', 'Paraguay', 'Benin', 'Montenegro', 'Mauritius', 'Lebanon', 'Macedonia, FYR', 'Uruguay', 'Maldives', 'Dominican Republic', 'Albania', 'Djibouti', 'Burkina Faso', 'Armenia', 'Honduras', 'Sri Lanka', 'Mali', 'Malawi', 'Uzbekistan', 'El Salvador', 'Uganda', 'Solomon Islands', 'Ghana', 'Comoros', 'Kyrgyz Republic', 'Zambia', 'Bahamas, The', 'Jamaica', 'Nepal', 'Congo, Rep.', 'Moldova', 'Algeria', 'Togo', 'Georgia', 'Sudan', 'Chad', 'Trinidad and Tobago', 'Guatemala', 'Turkmenistan', 'Gambia, The', 'Madagascar', 'Guinea-Bissau', 'Mongolia', 'Bolivia', 'Guyana', 'Fiji', 'Tajikistan', 'Mauritania', 'Equatorial Guinea', 'Syrian Arab Republic', 'Lesotho', 'Yemen, Rep.', 'Venezuela, RB', 'Liberia', 'Somalia', 'Guinea', 'Cuba', 'Iraq', 'Papua New Guinea', 'Bhutan', 'Gabon', 'Central African Republic', 'Zimbabwe', 'Haiti', 'Libya', 'Eritrea', 'Sierra Leone

c:\Users\Aditya\anaconda3\envs\wion\Lib\site-packages\openpyxl\worksheet\_read_only.py:85: UserWarning: Unknown extension is not supported and will be removed
  for idx, row in parser.parse():


Damped model ready: A(I-D) computed
  Mean diagonal of D : 0.3576

Shock   : max_V
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Damped  — nodes RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Delta   — mean absorption: 0.0000 | Max absorption: 0.0045 (TUN_O)
Damped* — common-ref RL>10%: 2 | Max RL: 1.0000 (TUN_D)
Delta*  — common-ref mean absorption: 0.0000 | Max absorption: 0.0045 (TUN_O)

Shock   : max_d
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 58 | Max RL: 1.0000 (DNK_H50)
Damped  — nodes RL>10%: 11 | Max RL: 1.0000 (DNK_H50)
Delta   — mean absorption: 0.0120 | Max absorption: 0.4399 (DNK_C301)
Damped* — common-ref RL>10%: 11 | Max RL: 1.0000 (DNK_H50)
Delta*  — common-ref mean absorption: 0.0120 | Max absorption: 0.4399 (DNK_C301)

Shock   : max_phi
Mode    : single_node | Nodes: 1
Baseline — nodes RL>10%: 1 | Max RL: 1.0000 (DEU_A01)
Damped  — nodes RL>10%: 1 | Max RL: 1.0000 (DEU_A01)
Delta   — mean absorption: 0.0020 | Max absorption: 

,label,mode,shock_nodes,mean_rl_base,mean_rl_damp,mean_delta_rl,median_delta_rl,max_delta_rl,top_country_delta,top_country_delta_value,country_delta_mean
0,max_V,single_node,1,0.000089,0.000065,0.000024,0.000006,0.004317,TUN,0.001444,0.000024
1,max_d,single_node,1,0.014260,0.002291,0.011968,0.006805,0.440476,DNK,0.144892,0.012070
2,max_phi,single_node,1,0.002156,0.000142,0.002014,0.001695,0.033586,DEU,0.009688,0.002021
3,RUS_C19,single_node,1,0.025330,0.006270,0.019059,0.008971,0.386605,RUS,0.280298,0.019076
4,USA_C19,single_node,1,0.016388,0.003279,0.013109,0.009755,0.200104,CAN,0.093545,0.013100


,scale,label,mode,shock_nodes,mean_rl_base,mean_rl_damp,mean_delta_rl,median_delta_rl,max_delta_rl,top_country_delta,top_country_delta_value,country_delta_mean
0,0.50,max_V,single_node,1,0.000089,0.000074,0.000014,0.000004,0.002458,TUN,0.000823,0.000015
1,0.50,max_d,single_node,1,0.014260,0.005883,0.008377,0.005135,0.221595,DNK,0.091034,0.008443
2,0.50,max_phi,single_node,1,0.002156,0.000508,0.001647,0.001391,0.020101,DEU,0.007191,0.001653
3,0.50,RUS_C19,single_node,1,0.025330,0.012224,0.013106,0.006521,0.236513,RUS,0.180794,0.013117
4,0.50,USA_C19,single_node,1,0.016388,0.006741,0.009647,0.007503,0.116789,CAN,0.065056,0.009642
5,0.75,max_V,single_node,1,0.000089,0.000069,0.000019,0.000005,0.003449,TUN,0.001154,0.000020
6,0.75,max_d,single_node,1,0.014260,0.003750,0.010509,0.006176,0.331297,DNK,0.121028,0.010595
7,0.75,max_phi,single_node,1,0.002156,0.000263,0.001893,0.001597,0.027315,DEU,0.008699,0.001899
8,0.75,RUS_C19,single_node,1,0.025330,0.008685,0.016644,0.008021,0.315025,RUS,0.237719,0.016659
9,0.75,USA_C19,single_node,1,0.016388,0.004601,0.011787,0.008955,0.161439,CAN,0.081942,0.011780


,label,n_boot,mean_delta_rl,std_delta_rl,p05_delta_rl,p95_delta_rl,mean_top_country_delta,mean_country_delta
0,RUS_C19,5,0.018769,3.946762e-04,0.018384,0.019196,0.277171,0.018786
1,USA_C19,5,0.013239,1.214688e-04,0.013116,0.013372,0.094987,0.013230
2,max_V,5,0.000024,8.365626e-07,0.000023,0.000024,0.001437,0.000024
3,max_d,5,0.012007,5.448233e-05,0.011936,0.012055,0.146242,0.012109
4,max_phi,5,0.002007,4.195977e-05,0.001952,0.002047,0.009670,0.002014


,scenario,mean_delta_rl,top_country,top_country_delta
0,max_V,0.000024,TUN,0.001444
1,max_d,0.011968,DNK,0.144892
2,max_phi,0.002014,DEU,0.009688
3,RUS_C19,0.019059,RUS,0.280298
4,USA_C19,0.013109,CAN,0.093545


## Presentation Notes

### Core message
The damped model is not a cosmetic variant of the baseline. It changes the propagation rule so damping acts on transmitted shock loss, which makes the comparison between baseline and damped outcomes economically interpretable.

### What the robustness checks show
- The scenario ranking is stable under alternative shock choices.
- Stronger damping lowers average delta_RL, while weaker damping moves results toward the baseline.
- Bootstrap perturbations show whether the headline findings survive small changes in the input-output structure.

### Caveats to state explicitly
- The model is linear and uses fixed technical coefficients.
- Damping is based on a structural/institutional proxy, not directly observed firm behavior.
- Results describe propagation under the modeled shocks, not full general equilibrium adjustment.

### Suggested slide order
1. Research question and why shock propagation matters
2. Baseline vs damped dynamic rule
3. Main results at node and country layers
4. Robustness: sensitivity sweep and bootstrap
5. Limitations and next validation steps

In [3]:
# Compact presentation export
scenario_block = scenario_summary.copy()
scenario_block['section'] = 'scenario'

bootstrap_block = bootstrap_summary.copy()
bootstrap_block['section'] = 'bootstrap'

presentation_export = pd.concat([scenario_block, bootstrap_block], ignore_index=True, sort=False)
presentation_export = presentation_export.reindex(columns=[
    'section', 'label', 'n_boot', 'mean_delta_rl', 'std_delta_rl',
    'p05_delta_rl', 'p95_delta_rl', 'mean_top_country_delta', 'mean_country_delta'
])

presentation_export.to_csv(output_dir / 'presentation_summary.csv', index=False)
display(presentation_export.head(10))
print('Saved presentation_summary.csv')

,section,label,n_boot,mean_delta_rl,std_delta_rl,p05_delta_rl,p95_delta_rl,mean_top_country_delta,mean_country_delta
0,scenario,max_V,NaN,0.000024,NaN,NaN,NaN,NaN,NaN
1,scenario,max_d,NaN,0.011968,NaN,NaN,NaN,NaN,NaN
2,scenario,max_phi,NaN,0.002014,NaN,NaN,NaN,NaN,NaN
3,scenario,RUS_C19,NaN,0.019059,NaN,NaN,NaN,NaN,NaN
4,scenario,USA_C19,NaN,0.013109,NaN,NaN,NaN,NaN,NaN
5,bootstrap,RUS_C19,5.0,0.018769,3.946762e-04,0.018384,0.019196,0.277171,0.018786
6,bootstrap,USA_C19,5.0,0.013239,1.214688e-04,0.013116,0.013372,0.094987,0.013230
7,bootstrap,max_V,5.0,0.000024,8.365626e-07,0.000023,0.000024,0.001437,0.000024
8,bootstrap,max_d,5.0,0.012007,5.448233e-05,0.011936,0.012055,0.146242,0.012109
9,bootstrap,max_phi,5.0,0.002007,4.195977e-05,0.001952,0.002047,0.009670,0.002014


Saved presentation_summary.csv


# Thesis Draft Map

Use this notebook as the working defense notebook and the thesis markdown files as the written chapter draft.

## Narrative flow
1. [Thesis index](../docs/thesis/00_thesis_index.md)
2. [Introduction and research question](../docs/thesis/01_introduction_and_question.md)
3. [Data and network construction](../docs/thesis/02_data_and_network.md)
4. [Model and method](../docs/thesis/03_model_and_method.md)
5. [Results and interpretation](../docs/thesis/04_results_and_interpretation.md)
6. [Robustness and validation](../docs/thesis/05_robustness_and_validation.md)
7. [Limitations and next steps](../docs/thesis/06_limitations_and_next_steps.md)
8. [Presentation outline](../docs/thesis/07_presentation_outline.md)

## What this section is for
- It gives a clean top-to-bottom thesis story.
- It keeps the notebook aligned with the written draft.
- It gives you a presentation-ready sequence you can lift into slides.

## Immediate next action
Run the notebook from the top once more after opening it, so the printed robustness summary and the exported presentation table are refreshed in the current kernel.

## Network Structure Effect: Hub-Dependent vs Scattered Nodes

This section analyzes how nodes with different upstream vulnerability profiles (V_j) experience different levels of shock absorption from damping. High-V_j nodes rely on concentrated suppliers (hubs); low-V_j nodes have more distributed supply chains.

In [5]:
import matplotlib.pyplot as plt
import seaborn as sns

# Extract V_j (upstream vulnerability) for all nodes
V = base['V'].copy()
nodes = V.index

# For each scenario result, compute delta_RL at the node level
# and stratify by V_j quintiles

vj_quintiles = pd.qcut(V, q=5, labels=['Q1_Low', 'Q2', 'Q3', 'Q4', 'Q5_High'], duplicates='drop')

stratified_results = []

for label, result in all_results.items():
    delta_rl = result['delta_RL'].copy().drop(index=result['shock_nodes'], errors='ignore')
    
    # Align with V_j quintiles
    delta_rl_df = pd.DataFrame({
        'delta_rl': delta_rl,
        'vj_quintile': vj_quintiles.reindex(delta_rl.index),
    }).dropna()
    
    # Stratified statistics
    for quintile in delta_rl_df['vj_quintile'].unique():
        group = delta_rl_df[delta_rl_df['vj_quintile'] == quintile]['delta_rl']
        stratified_results.append({
            'scenario': label,
            'vj_quintile': quintile,
            'mean_delta_rl': float(group.mean()),
            'median_delta_rl': float(group.median()),
            'std_delta_rl': float(group.std()),
            'n_nodes': len(group),
            'max_delta_rl': float(group.max()),
        })

stratified_df = pd.DataFrame(stratified_results)
stratified_df.to_csv(output_dir / 'vj_stratified_damping_effect.csv', index=False)

print('V_j Stratification: Mean delta_RL by quintile and scenario')
print(stratified_df.pivot_table(
    index='vj_quintile',
    columns='scenario',
    values='mean_delta_rl',
    aggfunc='first'
).round(6).to_string())

# Visualization: delta_RL across V_j quintiles
fig, axes = plt.subplots(1, 2, figsize=(14, 5))

# Panel 1: Mean delta_RL by V_j quintile across all scenarios
ax = axes[0]
pivot_mean = stratified_df.pivot_table(
    index='vj_quintile',
    columns='scenario',
    values='mean_delta_rl',
    aggfunc='first'
)
pivot_mean.plot(ax=ax, marker='o', linewidth=2)
ax.set_title('Mean delta_RL by Supply-Chain Vulnerability Quintile')
ax.set_xlabel('V_j Quintile (Low → High hub-dependence)')
ax.set_ylabel('Mean delta_RL (damping effect)')
ax.legend(title='Scenario', fontsize=9, loc='best')
ax.grid(True, alpha=0.3)

# Panel 2: Boxplot of max delta_RL by V_j quintile
ax = axes[1]
sns.boxplot(data=stratified_df, x='vj_quintile', y='max_delta_rl', ax=ax, color='#9ecae1')
ax.set_title('Distribution of Max delta_RL by V_j Quintile')
ax.set_xlabel('V_j Quintile (Low → High hub-dependence)')
ax.set_ylabel('Max delta_RL per node')
ax.grid(True, alpha=0.3, axis='y')

fig.tight_layout()
fig.savefig(figure_dir / 'vj_stratified_damping_effect.png', dpi=160, bbox_inches='tight')
plt.close(fig)

print(f'\nSaved vj_stratified_damping_effect.png and vj_stratified_damping_effect.csv')
print('\nInterpretation:')
print('- High V_j (Q5) nodes depend on concentrated suppliers.')
print('- Low V_j (Q1) nodes have dispersed, substitutable supply sources.')
print('- Comparing delta_RL across quintiles shows whether damping')
print('  is equally effective regardless of network position.')

V_j Stratification: Mean delta_RL by quintile and scenario
scenario      RUS_C19   USA_C19     max_V     max_d   max_phi
vj_quintile                                                  
Q1_Low       0.016985  0.012364  0.000018  0.011968  0.002124
Q2           0.020667  0.012706  0.000014  0.012435  0.001956
Q3           0.019802  0.013540  0.000010  0.012552  0.002140
Q4           0.018646  0.013985  0.000029  0.011510  0.002053
Q5_High      0.019198  0.012950  0.000047  0.011376  0.001796

Saved vj_stratified_damping_effect.png and vj_stratified_damping_effect.csv

Interpretation:
- High V_j (Q5) nodes depend on concentrated suppliers.
- Low V_j (Q1) nodes have dispersed, substitutable supply sources.
- Comparing delta_RL across quintiles shows whether damping
  is equally effective regardless of network position.


## Key Findings from V_j Stratification

**Network structure effect:**
The analysis above compares nodes grouped by upstream vulnerability (V_j) to see whether hub-dependent and scattered-supply nodes experience different damping benefits.

**What we observe:**
- Mean delta_RL is relatively **flat across V_j quintiles** (ranges 0.012–0.021 for main shocks).
- This suggests that the **damping effect is robust to network position**, not just benefiting high-V_j nodes.
- Nodes with dispersed suppliers (low V_j, Q1) still absorb shock via damping, even though they have substitutes.
- Nodes with concentrated suppliers (high V_j, Q5) also benefit from damping, even though substitution is harder.

**Interpretation:**
The shock-only damping model works through institutional capacity and structural resilience, both of which apply across the network. A low-V_j node in a high-LPI country can still absorb loss via institutional efficiency. A high-V_j node in any country benefits from the multiplicative effect of (1 - V_j) × φ_j.

**For the thesis:**
This finding strengthens the defensibility of the damped model. The damping benefit is not an artifact of specific network positions; it represents a systematic, network-agnostic mechanism for loss absorption tied to economic structure and institutional capacity.